### PR TITLE
Implement Right Join in NestedLoopOperation

### PR DESCRIPTION
**Pull request by matriv**

### CODE REVIEW

File: sql/src/main/java/io/crate/operation/join/NestedLoopOperation.java [hunk 1]
```diff
@@ -408,7 +418,7 @@ private Result emitRow(Row row) {
 
         @Override
         public void finish(final RepeatHandle repeatHandle) {
-            if (joinType == JoinType.LEFT && !matched && left.lastRow != null) {
+            if (JoinType.LEFT.equals(joinType) && !matched && left.lastRow != null) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/join/NestedLoopOperation.java]: why the change to equals? == is fine on enum?

File: sql/src/test/java/io/crate/operation/join/NestedLoopOperationTest.java [hunk 2]
```diff
@@ -327,21 +328,42 @@ public void run() {
...
+                                                         "4| NULL\n" +
+                                                         "5| 5\n"));
+    }
+
+    @Test
+    public void testNestedLoopOperationWithRightOuterJoin() throws Exception {
+        List<Row> leftRows = asRows(3, 5);
+        List<Row> rightRows = asRows(1, 2, 3, 4, 5);
+
+        Predicate<Row> rowFilterPredicate = new Predicate<Row>() {
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/operation/join/NestedLoopOperationTest.java]: could this be a attribute with is re-used? another test uses the same predicate

File: sql/src/main/java/io/crate/operation/join/BitSet.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+ * agreement.
+ */
+
+package io.crate.operation.join;
+
+import org.apache.lucene.util.LongBitSet;
+
+import java.util.ArrayList;
+
+class BitSet {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/join/BitSet.java]: can we have some docs here explaining this class? The difference to `java.util.BitSet` etc. would be interesting

File: sql/src/main/java/io/crate/operation/join/NestedLoopOperation.java [hunk 4]
```diff
@@ -512,9 +522,64 @@ public void resume(boolean async) {
...
+        private final BitSet matchedRows = new BitSet();
+        private long rowPosition = -1;
+
+        RightJoinRightRowReceiver(int leftNumOutputs, int rightNumOutputs) {
+            super(leftNumOutputs, rightNumOutputs);
+        }
+
+        @Override
+        public Result setNextRow(final Row rightRow) {
+            if (!stop) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/join/NestedLoopOperation.java]: why does this still do something if stop is true?

Comment (matriv) [inline sql/src/main/java/io/crate/operation/join/NestedLoopOperation.java]: I want to just increase the rowPosition if !stop and avoid copying code from superclass.

File: sql/src/main/java/io/crate/operation/join/NestedLoopOperation.java [hunk 5]
```diff
@@ -512,9 +522,64 @@ public void resume(boolean async) {
...
+                    combinedRow.outerRow = combinedRow.outerNullRow;
+                    combinedRow.innerRow = row;
+                    return emitRowAndTrace(combinedRow);
+                }
+                return Result.CONTINUE;
+            } else {
+                combinedRow.outerRow = left.lastRow;
+                combinedRow.innerRow = row;
+
+                // filter logic
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/join/NestedLoopOperation.java]: Not sure if this comment here is helpful

**Comment by mfussenegger** [comment]
I think the docs on NestedLoopOperation should be extended to explain how this new algorithm works

File: sql/src/main/java/io/crate/operation/join/BitSet.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+import java.util.ArrayList;
+
+/**
+ * This BitSet is used to mark matched rows between left and right in {@link NestedLoopOperation}
+ *
+ * Each bit true if the rows in the respective position are matched and therefore we need
+ * a structure capable of holding <pre>long</pre> size of bits so java.util.BitSet cannot be used.
+ *
+ * We choose to use {@link LongBitSet} from Lucence and add another layer of segments on top in
+ * order to further optimize performance. The usage of segments prevents from calling System.arrayCopy()
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/join/BitSet.java]: are there any proofs for the optimized performance?

File: sql/src/main/java/io/crate/operation/join/BitSet.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+ * and instead when the current segments are full we reserve a new segment and add it to the segments
+ * list.
+ */
+class BitSet {
+
+    private static long SEGMENT_SIZE = 1048576;
+    private static int INITIAL_NO_OF_SEGMENTS = 10;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/join/BitSet.java]: Lucene docs says LongBitSet should be used over FixedBitSet if more than 2.1B bits are stored, but we limit each bitset to only 1048576 but then initiate 10 of them? 
Does that really make sense?

File: sql/src/main/java/io/crate/operation/join/NestedLoopOperation.java [hunk 8]
```diff
@@ -68,6 +68,10 @@
  *             RowReceiver
  * </pre>
  *
+ * In case of Right & Full (Outer) Join once the whole left-right loop is finished a last loop
+ * on the right is executed in order to emit rows from the right that weren't matched with rows
+ * from left.
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/join/NestedLoopOperation.java]: Maybe also include this as a pseudo code example.. eg.

```
for (leftRow in left)
    for (rightRow in right)
        saveIfMatched(position)
        emitIfMatch
for (rightRow in right)
    if notMatched:
        emitWithLeftAsNull
```

Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/join/NestedLoopOperation.java]: So far the NestedLoop also kept the order of the Rows it received. With Right join this is no longer the case - I think that should be mentioned as well

File: sql/src/main/java/io/crate/operation/join/BitSet.java (hunk 7 above)
Comment (matriv) [inline sql/src/main/java/io/crate/operation/join/BitSet.java]: FixedBitSet has get(int index) but we discussed with Bernd that we want long to be able to manage large data sets on the right side. Furthermore we need to be able to dynamically grow since we don't know the size beforehand. The purpose of starting with 10 segments of 1024^2 bits is not to reserve too much memory in the beginning but still be able to grow dynamically. We could directly use the LongBitSet and the ensureCapacity but this makes a copy of all the existing bits in order to grow, where in our case we just grow the List of segments which holds references to the fixed underlying LongBitSets.

File: sql/src/main/java/io/crate/operation/join/NestedLoopOperation.java [hunk 9]
```diff
@@ -512,9 +524,64 @@ public void resume(boolean async) {
...
+            }
+            return super.setNextRow(rightRow);
+        }
+
+        protected Result emitRow(Row row) {
+            if (emmitRightJoin) {
+                if (!matchedRows.get(rowPosition)) {
+                    combinedRow.outerRow = combinedRow.outerNullRow;
+                    combinedRow.innerRow = row;
+                    return emitRowAndTrace(combinedRow);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/join/NestedLoopOperation.java]: further below the pauseFromDownstream flag is set, but here it isn't?

**Comment by mfussenegger** [comment]
lgtm
